File: Sean的leetcode/Leetcode/Word Break.cpp
//refer http://fisherlei.blogspot.tw/2013/11/leetcode-word-break-solution.html
//Time complexity:O(n^2)
class Solution {
public:
    bool wordBreak(string s, unordered_set<string>& wordDict) {
        int n = s.size();
        if (n == 0) return false;
        vector<bool> dp(n+1, false);

        dp[0] = true;
        //dp[i] is true means s.substr(0, i) (length is i) is a valid word Break.
        for (int i = 1; i <= n; i++)
            for (int j = 0; j < i; j++)
            {
                if (dp[j] && wordDict.count(s.substr(j, i-j))) {
                    dp[i] = true;
                    break;
                }
            }
        return dp[n];
    }
};

File: Sean的leetcode/FB/Word Break.py
###
# DP
# Time Complexit: O(n^2)
# Space Complexity: O(n)
###
class Solution(object):
    def wordBreak(self, s, wordDict):
        """
        :type s: str
        :type wordDict: List[str]
        :rtype: bool
        """
        if not s or not wordDict:
            return False
        dp = [False] * (len(s) + 1)
        dp[0] = True
        for i in xrange(1, len(s) + 1):
            for j in xrange(i):
                if dp[j] and s[j:i] in wordDict:
                    dp[i] = True
                    break
        return dp[-1]